### PR TITLE
Coded features review: LP fees & rewards #4811

### PR DESCRIPTION
**Pull request by Jiajia-Cui**
Reviewing the feature test in issue #4814 [https://github.com/vegaprotocol/vega/issues/4814]

Initialised  from original test and added 2 scenarios 0042-LP-Fees-004 and 0042-LP-Fees-005 to cover edge cases : https://github.com/vegaprotocol/specs-internal/blob/lp-fees-test/qa-scenarios/LP-Fees-Rewards-Test.feature with added Acceptance criteria number

Create an extra case to cover scenario 0042-LP-Fees-006 when distribution period is very large: https://github.com/vegaprotocol/specs-internal/blob/lp-fees-test/qa-scenarios/LP-Fees-Rewards-Large-Distribution-Time.feature

### CODE REVIEW

File: qa-scenarios/Debug-Featuretest-BondSlashing.json [hunk 1]
```diff
@@ -0,0 +1,33 @@
+{
```
Comment (witgaw) [inline qa-scenarios/Debug-Featuretest-BondSlashing.json:1]: This file shouldn't be committed.

File: protocol/0034-PROB-prob_weighted_liquidity_measure.ipynb [hunk 2]
```diff
@@ -72,10 +72,10 @@
     "\n",
     "## Acceptance Criteria\n",
     "\n",
-    "1. It is always true that for $x_{min}$, $x_{max}$ (from risk model) and $S$ we work with (either `mid` or `last trade price` as appropriate) we have $x_{min} < S < x_{max}$. (<a name=\"0034-PROB-001\" href=\"#0034-PROB-001\">0034-PROB-001</a>)\n",
-    "2. If there is volume on either side of the order book within $[x_{min},x_{max}]$ then the reported liquidity is strictly more than $0$ when using the `log-normal` risk model (with any valid parameters). (<a name=\"0034-PROB-002\" href=\"#0034-PROB-002\">0034-PROB-002</a>)\n",
-    "3. If there is no mid price in continuous trading then the reported liquidity is $0$. (<a name=\"0034-PROB-003\" href=\"#0034-PROB-003\">0034-PROB-003</a>)\n",
-    "4. If there is no last trade price in an auction (opening auction in particular) then the reported liquidity is $0$. (<a name=\"0034-PROB-004\" href=\"#0034-PROB-004\">0034-PROB-004</a>)\n"
+       "1. Pegged Price and Volume from LP commitment should carry the correct amount\n",
```
Comment (witgaw) [inline protocol/0034-PROB-prob_weighted_liquidity_measure.ipynb:81]: I think tags are needed (e.g. `(<a name=\"0034-PROB-001\" href=\"#0034-PROB-001\">0034-PROB-001</a>)`).

File: protocol/0034-PROB-prob_weighted_liquidity_measure.ipynb [hunk 3]
```diff
@@ -72,10 +72,10 @@
...
+    "2. If there is volume on either side of the order book within $[x_{min},x_{max}]$ then the reported liquidity is strictly more than $0$ when using the `log-normal` risk model (with any valid parameters)\n",
+    "3. More details please see features tests: integration/features/liquidity-provisin \n",
```
Comment (witgaw) [inline protocol/0034-PROB-prob_weighted_liquidity_measure.ipynb:83]: This is a bit circular, ACs are meant to define what behaviours are expected and tests need to confirm that it's the case (they can still do other things of course). 

Perhaps we could substitute it with some concrete ones like: probability decreases away from the best bid ask price for buy orders and from the best ask price for sell orders.

File: qa-scenarios/LP-Fees-Rewards-Large-Distribution-Time.feature [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+      | party2 | ETH/MAR22 | sell | 1      | 1100  | 0                | TYPE_LIMIT | TIF_GTC |
+      | party2 | ETH/MAR22 | sell | 10     | 1000  | 0                | TYPE_LIMIT | TIF_GTC |
+
+    Then the opening auction period ends for market "ETH/MAR22"
+
+    And the following trades should be executed:
+      | buyer   | price | size | seller |
+      | party1  | 1000  | 10   | party2 |
+
+    Then debug transfers
```
Comment (witgaw) [inline qa-scenarios/LP-Fees-Rewards-Large-Distribution-Time.feature]: Nitpick: if possible we should remove any debug statements before merging the tests.

File: qa-scenarios/LP-Fees-Rewards-Large-Distribution-Time.feature [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+      | buyer   | price | size | seller |
+      | party1  | 951   | 15   | lp1    |
+
+    # lp fee got cumulated since the distribution period is large
+    And the accumulated liquidity fees should be "35" for the market "ETH/MAR22"
+
+    # lp fee got cumulated since the distribution period is large
+    Then time is updated to "2020-11-30T00:30:05Z"
+
+    And the accumulated liquidity fees should be "35" for the market "ETH/MAR22"
```
Comment (witgaw) [inline qa-scenarios/LP-Fees-Rewards-Large-Distribution-Time.feature]: I think it would be good to finish this off by moving time forward so that accumulated liquidity fees get paid out to lp1 (checking transfers and balances).

File: qa-scenarios/LP-Fees-Rewards-Large-Distribution-Time.feature [hunk 6]
```diff
@@ -0,0 +1,154 @@
+Feature: Test liquidity provider reward distribution; Check what happens when distribution period is large(both in genesis and mid-market)
```
Comment (witgaw) [inline qa-scenarios/LP-Fees-Rewards-Large-Distribution-Time.feature]: The "mid-market" bit never seems to be checked, but I'm not sure how important that is.

File: qa-scenarios/LP-Fees-Rewards-Test.feature [hunk 7]
```diff
@@ -0,0 +1,544 @@
...
+      | buyer   | price | size | seller  |
+      | party2 | 951   | 12   | lp1     |
+      | party2 | 1000  | 8    | party1 |
+
+    And the accumulated liquidity fees should be "20" for the market "ETH/MAR22"
+
+    # opening auction + time window
+    Then time is updated to "2019-11-30T00:10:05Z"
+
+    # these are different from the tests, but again, we end up with a 2/3 vs 1/3 fee share here.
```
Comment (witgaw) [inline qa-scenarios/LP-Fees-Rewards-Test.feature]: I think it's 0.8 vs 0.2 (as implied by equity-like share).

File: qa-scenarios/LP-Fees-Rewards-Test.feature [hunk 8]
```diff
@@ -0,0 +1,544 @@
...
+      | buyer   | price | size | seller |
+      | party1 | 951   | 12   | lp1    |
+      | party1 | 951   | 3    | lp2    |
+
+    And the accumulated liquidity fees should be "15" for the market "ETH/MAR22"
+
+    # opening auction + time window
+    Then time is updated to "2019-11-30T00:20:06Z"
+
+    # these are different from the tests, but again, we end up with a 2/3 vs 1/3 fee share here.
```
Comment (witgaw) [inline qa-scenarios/LP-Fees-Rewards-Test.feature]: Same as above.

File: qa-scenarios/LP-Fees-Rewards-Test.feature [hunk 9]
```diff
@@ -0,0 +1,544 @@
...
+
+    # no fees in auction
+    And the accumulated liquidity fees should be "0" for the market "ETH/MAR22"
+
+    Then the parties place the following orders:
+      | party  | market id | side | volume | price | resulting trades | type       | tif     | reference   |
+      | party1 | ETH/MAR22 | sell | 20     | 1000  | 0                | TYPE_LIMIT | TIF_GTC | party1-sell |
+      | party2 | ETH/MAR22 | buy  | 20     | 1000  | 1                | TYPE_LIMIT | TIF_GTC | party2-buy  |
+
+    And the accumulated liquidity fees should be "20" for the market "ETH/MAR22"
```
Comment (witgaw) [inline qa-scenarios/LP-Fees-Rewards-Test.feature:424]: I think it would be worth to verify the fee distribution.

File: qa-scenarios/LP-Fees-Rewards-Test.feature [hunk 10]
```diff
@@ -0,0 +1,544 @@
...
+
+  Scenario: 2 LPs joining at start, unequal commitments, Check what happens when distribution period gets set to a larger value (0042-LP-Fees-005)
```
Comment (witgaw) [inline qa-scenarios/LP-Fees-Rewards-Test.feature]: Unless I'm missing something this test doesn't seem to do what the description says.

File: qa-scenarios/liquidity-provision-bond-account.feature [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+      | party0 | ETH/MAR22 | sell | 15     | 1000  | 0                | TYPE_LIMIT | TIF_GTC | party0-sell-3  |
+
+  Then debug transfers  
+  Then debug orders
+
+  And the market data for the market "ETH/MAR22" should be:
+      | mark price | trading mode            | horizon | min bound | max bound | target stake | supplied stake | open interest |
+      | 1000       | TRADING_MODE_CONTINUOUS | 1       | 1000      | 1000      | 213414       | 50000          | 60            |
+
+  And the insurance pool balance should be "8154" for the market "ETH/MAR22"
```
Comment (witgaw) [inline qa-scenarios/liquidity-provision-bond-account.feature]: Might be worth reconciling this figure with the shortfall (check MTM transfers) to make sure that it's indeed 0.2 (penalty factor) of the shortfall amount.

File: qa-scenarios/bond-slashing.feature.mock [hunk 12]
```diff
@@ -1,27 +1,128 @@
-Feature: Test liquidity provider bond slashing
```
Comment (witgaw) [inline qa-scenarios/bond-slashing.feature.mock]: Is this still meant to be merged as a .mock?

**Review by witgaw — CHANGES_REQUESTED** [review]
Looks really good overall, please see comments below before merging though.

File: qa-scenarios/bond-slashing.feature.mock (hunk 12 above)
Comment (Jiajia-Cui) [reply qa-scenarios/bond-slashing.feature.mock]: no, it should be removed. :)

File: qa-scenarios/LP-Fees-Rewards-Test.feature (hunk 10 above)
Comment (Jiajia-Cui) [reply qa-scenarios/LP-Fees-Rewards-Test.feature]: updated the description

File: qa-scenarios/liquidity-provision-bond-account.feature (hunk 11 above)
Comment (Jiajia-Cui) [reply qa-scenarios/liquidity-provision-bond-account.feature]: I am still looking at this one, I think there might an issue, but Elias thinks that its not a defect, in this test case bond got penalised for having bond account empty and then for having margin and general account empty. Still doing some investigation at the moment. 

https://github.com/vegaprotocol/vega/issues/4627

File: qa-scenarios/bond-slashing.feature.mock (hunk 12 above)
Comment (Jiajia-Cui) [reply qa-scenarios/bond-slashing.feature.mock]: and the feature test should replace the mock one

File: qa-scenarios/liquidity-provision-bond-account.feature (hunk 11 above)
Comment (Jiajia-Cui) [reply qa-scenarios/liquidity-provision-bond-account.feature]: It's a behaviour when LP provider added a new order, network tries to top up the margin account with the additional order, and then when margin account is below Margin Release level, the margin account stays. So not a defect.
